test/common/upstream/health_check_fuzz.cc

```diff
@@ -112,34 +111,402 @@ void HealthCheckFuzz::triggerTimeoutTimer(bool last_action) {
+  // set by multiple code paths. handleFailure() turns on interval and turns off timeout. However,
+  // other action of the fuzzer account for this by explicitly invoking a client after
+  // expect_close_ gets set to true, turning expect_close_ back to false.
+  connection_->raiseEvent(event_type);
+  if (!last_action && event_type != Network::ConnectionEvent::Connected) {
+    if (!interval_timer_->enabled_) {
+      return;
+    }
+    ENVOY_LOG_MISC(trace, "Creating client from close event");
+    expectClientCreate();
+    interval_timer_->invokeCallback();
+  }
+
+  // In the specific case of:
+  // https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/health_checker_impl.cc#L489
+  // This blows away client, should create a new one
+  if (event_type == Network::ConnectionEvent::Connected && empty_response_) {
+    ENVOY_LOG_MISC(trace, "Creating client from connected event and empty response.");
+    expectClientCreate();
+    interval_timer_->invokeCallback();
+  }
+}
+
+void GrpcHealthCheckFuzz::allocGrpcHealthCheckerFromProto(
+    const envoy::config::core::v3::HealthCheck& config) {
+  health_checker_ = std::make_shared<TestGrpcHealthCheckerImpl>(
+      *cluster_, config, dispatcher_, runtime_, random_,
+      HealthCheckEventLoggerPtr(event_logger_storage_.release()));
+  ENVOY_LOG_MISC(trace, "Created Test Grpc Health Checker");
+}
+
+void GrpcHealthCheckFuzz::initialize(test::common::upstream::HealthCheckTestCase input) {
+  allocGrpcHealthCheckerFromProto(input.health_check_config());
+  cluster_->prioritySet().getMockHostSet(0)->hosts_ = {
+      makeTestHost(cluster_->info_, "tcp://127.0.0.1:80")};
+  expectSessionCreate();
+  expectStreamCreate(0);
+  health_checker_->start();
+  ON_CALL(runtime_.snapshot_, getInteger("health_check.min_interval", _))
+      .WillByDefault(testing::Return(45000));
+
+  if (DurationUtil::durationToMilliseconds(input.health_check_config().initial_jitter()) != 0) {
+    test_sessions_[0]->interval_timer_->invokeCallback();
+  }
+  if (input.health_check_config().has_reuse_connection()) {
+    reuse_connection_ = input.health_check_config().reuse_connection().value();
+  }
+}
+
+// From unit tests
+std::vector<uint8_t> GrpcHealthCheckFuzz::serializeResponse(
+    grpc::health::v1::HealthCheckResponse::ServingStatus status) {
+  grpc::health::v1::HealthCheckResponse response;
+  response.set_status(status);
+  const auto data = Grpc::Common::serializeToGrpcFrame(response);
+  auto ret = std::vector<uint8_t>(data->length(), 0);
+  data->copyOut(0, data->length(), &ret[0]);
+  return ret;
+}
+
+Buffer::OwnedImpl GrpcHealthCheckFuzz::makeBufferToRespondWith(
+    test::common::upstream::GrpcRespondBytes grpc_respond_bytes) {
+  switch (grpc_respond_bytes.grpc_respond_bytes_selector_case()) {
+  case test::common::upstream::GrpcRespondBytes::kStatus: {
+    // Structured Response
+    std::vector<uint8_t> data;
+    switch (grpc_respond_bytes.status()) {
```

Comment:
I added the function, but just wondering, why is it better to put it in an anonymous namespace rather than just have it as a method in the grpc health check class?

Reply:
I think it's a little bit of preference, the style guide only mentions

> Functions intended to be local to a cc file should be declared in an anonymous namespace, rather than using the 'static' keyword. Note that the Google C++ style guide allows either, but in Envoy we prefer anonymous namespaces.

but basically if a function is only intended to be used inside the source code as a helper and no one else benefits, there's no need to clutter the API and header with it (this relates to another protobuf to type converter you had earlier, but I didn't make a strong case against it once it's there). It's a little up to preference!

Reply:
Hmmmm okay I will switch all the helper functions across all three health checkers to an anonymous namespace!

Reply:
Moved helper functions to anonymous namespace!